src/infra/ingest-datafusion/src/writer.rs

```diff
@@ -721,98 +721,117 @@ impl DataWriter for DataWriterDataFusion {
             (add_data, None, None)
         };
 
-        // Do we have anything to commit?
-        if add_data.new_offset_interval.is_none()
-            && add_data.new_watermark == self.meta.prev_watermark
-            && add_data.new_source_state == self.meta.prev_source_state
+        // Do we need to commit `SetDataSchema` event?
+        let new_schema = if self.meta.schema.is_none() {
+            new_schema
+        } else {
+            None
+        };
+
+        // Do we have anything to commit in `AddData` event?
+        let add_data = if add_data.new_offset_interval.is_some()
+            || add_data.new_watermark != self.meta.prev_watermark
+            || add_data.new_source_state != self.meta.prev_source_state
         {
-            Err(EmptyCommitError {}.into())
+            Some(add_data)
         } else {
+            None
+        };
+
+        if new_schema.is_some() || add_data.is_some() {
             Ok(StageDataResult {
                 system_time: opts.system_time,
                 add_data,
                 new_schema,
                 data_file,
             })
+        } else {
+            Err(EmptyCommitError {}.into())
         }
     }
 
     #[tracing::instrument(level = "info", skip_all)]
     async fn commit(&mut self, staged: StageDataResult) -> Result<WriteDataResult, CommitError> {
+        assert!(staged.new_schema.is_some() || staged.add_data.is_some());
+
         let old_head = self.meta.head.clone();
 
-        // Commit schema if it was not previously defined
-        if self.meta.schema.is_none() {
-            if let Some(new_schema) = staged.new_schema {
-                // TODO: Make commit of schema and data atomic
-                let commit_schema_result = self
-                    .dataset
-                    .commit_event(
-                        odf::SetDataSchema::new(&new_schema).into(),
-                        CommitOpts {
-                            block_ref: &self.block_ref,
-                            system_time: Some(staged.system_time),
-                            prev_block_hash: Some(Some(&self.meta.head)),
-                            check_object_refs: false,
-                            update_block_ref: true,
-                        },
-                    )
-                    .await?;
+        // Commit schema changes if any
+        if let Some(new_schema) = staged.new_schema {
```

Comment:
I don't think I understood what is the scenario when schema evolves suddenly?
Previously we discussed this is a complex topic of dataset data part migrations..

Reply:
This is not about schema evolution yet (will clarify the comment).

This is purely about writing `SetDataSchema` event even if we don't have anything to write into `AddData` yet for datasets that stay empty for a long time.